File: mobile/oncall_mobile/lib/models/payment_attempt.dart
import '../core/json.dart';

class PaymentAttempt {
  PaymentAttempt({
    required this.id,
    required this.method,
    required this.status,
    required this.gateway,
    this.gatewayReference,
    this.rejectionReason,
    this.createdAt,
  });

  factory PaymentAttempt.fromJson(Map<String, dynamic> json) =>
      PaymentAttempt(
        id: json['id'] as int,
        method: json['method'] as String,
        status: json['status'] as String,
        gateway: json['gateway'] as String,
        gatewayReference: json['gateway_reference'] as String?,
        rejectionReason: json['rejection_reason'] as String?,
        createdAt: parseDate(json['created_at']),
      );

  final int id;
  final String method;
  final String status;
  final String gateway;
  final String? gatewayReference;
  final String? rejectionReason;
  final DateTime? createdAt;
}

File: app/Models/PaymentAttempt.php
<?php

namespace App\Models;

use App\Enums\PaymentAttemptStatus;
use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

#[Fillable(['job_payment_id', 'method', 'status', 'gateway', 'gateway_reference', 'idempotency_key', 'submitted_by', 'reviewed_by', 'reviewed_at', 'rejection_reason', 'expires_at'])]
class PaymentAttempt extends Model
{
    protected function casts(): array
    {
        return [
            'status' => PaymentAttemptStatus::class,
            'reviewed_at' => 'datetime',
            'expires_at' => 'datetime',
        ];
    }

    public function jobPayment(): BelongsTo
    {
        return $this->belongsTo(JobPayment::class);
    }

    public function submitter(): BelongsTo
    {
        return $this->belongsTo(User::class, 'submitted_by');
    }

    public function reviewer(): BelongsTo
    {
        return $this->belongsTo(User::class, 'reviewed_by');
    }
}
